docs/rfc/rfc-007-deterministic-proto-bytes.md

```diff
@@ -0,0 +1,104 @@
+# RFC 008 : Deterministic Proto Byte Serialization
+
+## Changelog
+
+- 09-Dec-2021: Initial draft (@williambanfield).
+
+## Abstract
+
+This document discusses the issue of stable byte-representation of serialized messages
+within Tendermint and describes a few possible routes that could be taken to address it.
+
+## Background
+
+We use the byte representations of wire-format proto messages to produce
+and verify hashes of data within the Tendermint codebase as well as for
+producing and verifying cryptographic signatures over these signed bytes.
+
+The protocol buffer [encoding spec][proto-spec-encoding] does not guarantee that the byte representation
+of a protocol buffer message will be the same between two calls to an encoder.
+While there is a mode to force the encoder to produce the same byte representation
+of messages within a single binary, these guarantees are not good enough for our
+use case in Tendermint. We require multiple different versions of a binary running
+Tendermint to be able to inter-operate. Additionally, we require that multiple different
+systems written in _different languages_ be able to participate in different aspects
+of the protocols of Tendermint and be able to verify the integrity of the messages
+they each produce.
+
+While this has not yet created a problem that we know of in a running network, we should
+make sure to provide stronger guarantees around the serialized representation of the messages
+used within the Tendermint consensus algorithm to prevent any issue from occurring.
+
+
+## Discussion
+
+Proto has the following points of variability that can produce non-deterministic byte representation:
+
+1. Encoding order of fields within a message.
+
+Proto allows fields to be encoded in any order and even be repeated.
+
+2. Encoding order of elements of a repeated field.
+
+`repeated` fields in a proto message can be serialized in any order.
+
+3. Presence or absence of default values.
+
+Types in proto have defined default values similar to Go's zero values. 
+Writing or omitting a default value are both legal ways of encoding a wire message.
+
+4. Serialization of 'unknown' fields. 
+
+Unknown fields can be present when a message is created by a binary with a newer 
+version of the proto that contains fields that the deserializer in a different 
+binary does not yet know about. Deserializers in binaries that do not know about the field
+will maintain the bytes of the unknown field but not place them into the deserialized structure.
+
+We have a few options to consider when producing this stable representation.
+
+### Use only compliant serializers and constrain field usage
+
+According to [Cosmos-SDK ADR-27][cosmos-sdk-adr-27], when message types obey a simple 
+set of rules, gogoproto produces a consistent byte representation of serialized messages.
+This seems promising, although more research is needed to guarantee gogoproto always
```

Comment:
Don't we also have the intention of removing gogoproto?

Reply:
I believe that there is interest in doing this in the SDK, although I'm not sure I fully understand the timeline. It's likely a reasonable thing to do seeing as gogoproto is currently in maintenance limbo. However, in terms of concrete plans to remove it from Tendermint completely, I'm not aware of any at the moment.